dimod/include/dimod/constraint.h

```diff
@@ -0,0 +1,178 @@
+// Copyright 2022 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+#include <limits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "dimod/abc.h"
+#include "dimod/expression.h"
+#include "dimod/vartypes.h"
+
+namespace dimod {
+
+enum Sense { LE, GE, EQ };
+
+enum Penalty { LINEAR, QUADRATIC, CONSTANT };
+
+template <class Bias, class Index>
+class Constraint : public Expression<Bias, Index> {
+ public:
+    /// The type of the base class.
+    using base_type = Expression<Bias, Index>;
+
+    /// The first template parameter (Bias).
+    using bias_type = Bias;
+
+    /// The second template parameter (Index).
+    using index_type = Index;
+
+    using size_type = typename base_type::size_type;
+
+    using parent_type = ConstrainedQuadraticModel<bias_type, index_type>;
+
+    Constraint();
+    explicit Constraint(parent_type* parent);
+
+    bool is_onehot() const;
+    bool is_soft() const;
+    void mark_discrete(bool mark = true);
+    bool marked_discrete() const;
+    Penalty penalty() const;
+    bias_type rhs() const;
+
+    // note: flips sign when negative
+    void scale(bias_type scalar);
+
+    Sense sense() const;
+    void set_penalty(Penalty penalty);
+    void set_rhs(bias_type rhs);
+    void set_sense(Sense sense);
+    void set_weight(bias_type weight);
+    bias_type weight() const;
+
+ private:
+    Sense sense_;
+    bias_type rhs_;
+    bias_type weight_;
+    Penalty penalty_;
+
+    // marker(s) - these ar not enforced by code
+    bool marked_discrete_ = false;
+};
+
+template <class bias_type, class index_type>
+Constraint<bias_type, index_type>::Constraint() : Constraint(nullptr) {}
+
+template <class bias_type, class index_type>
+Constraint<bias_type, index_type>::Constraint(parent_type* parent)
+        : base_type(parent),
+          sense_(Sense::EQ),
+          rhs_(0),
+          weight_(std::numeric_limits<bias_type>::infinity()),
```

Comment:
This infinite weight could produce weird values if somehow it remains till the end and this constraint is marked soft. But a smaller value would also allow bugs to be undetected.

Reply:
Yup, would much prefer this method and let any potential solver test whether it knows how to handle it.